src/core/backbone_router/bbr_manager.cpp

```diff
@@ -98,61 +101,104 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
+
+    uint32_t timeout;
+    uint16_t commissionerSessionId;
+    bool     hasCommissionerSessionIdTlv = false;
+    bool     hasTimeoutTlv               = false;
+
     VerifyOrExit(aMessage.IsConfirmable() && aMessage.GetCode() == OT_COAP_CODE_POST, error = OT_ERROR_PARSE);
 
     VerifyOrExit(isPrimary, status = ThreadStatusTlv::kMlrBbrNotPrimary);
 
     // TODO: (MLR) send configured MLR response for Reference Device
-    // TODO: (MLR) handle Commissioner Session TLV
-    // TODO: (MLR) handle Timeout TLV
+
+    if (ThreadTlv::FindUint16Tlv(aMessage, ThreadTlv::kCommissionerSessionId, commissionerSessionId) == OT_ERROR_NONE)
+    {
+        const MeshCoP::CommissionerSessionIdTlv *commissionerSessionIdTlv =
+            static_cast<const MeshCoP::CommissionerSessionIdTlv *>(
+                Get<NetworkData::Leader>().GetCommissioningDataSubTlv(MeshCoP::Tlv::kCommissionerSessionId));
+
+        VerifyOrExit(commissionerSessionIdTlv != nullptr &&
+                         commissionerSessionIdTlv->GetCommissionerSessionId() == commissionerSessionId,
+                     error = OT_ERROR_PARSE);
+
+        hasCommissionerSessionIdTlv = true;
+    }
+
+    hasTimeoutTlv = hasCommissionerSessionIdTlv &&
+                    (ThreadTlv::FindUint32Tlv(aMessage, ThreadTlv::kTimeout, timeout) == OT_ERROR_NONE);
 
     VerifyOrExit(ThreadTlv::FindTlvValueOffset(aMessage, IPv6AddressesTlv::kIPv6Addresses, addressesOffset,
                                                addressesLength) == OT_ERROR_NONE,
                  error = OT_ERROR_PARSE);
     VerifyOrExit(addressesLength % sizeof(Ip6::Address) == 0, error = OT_ERROR_PARSE);
     VerifyOrExit(addressesLength / sizeof(Ip6::Address) <= kIPv6AddressesNumMax, error = OT_ERROR_PARSE);
 
-    SuccessOrExit(error = Get<BackboneRouter::Leader>().GetConfig(config));
-    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(config.mMlrTimeout);
+    if (!hasTimeoutTlv)
+    {
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetConfig(config));
 
-    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+        timeout = config.mMlrTimeout;
+    }
+    else
     {
-        bool failed = true;
+        VerifyOrExit(timeout < std::numeric_limits<uint32_t>::max(), status = ThreadStatusTlv::kMlrNoPersistent);
+
+        if (timeout != 0)
+        {
+            timeout = std::min(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMax));
+            timeout = std::max(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMin));
+        }
+    }
+
+    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(timeout);
 
+    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+    {
         IgnoreReturnValue(aMessage.Read(addressesOffset + offset, sizeof(Ip6::Address), &address) ==
                           sizeof(Ip6::Address));
 
-        switch (mMulticastListenersTable.Add(address, expireTime))
+        if (timeout == 0)
```

Comment:
```suggestion
        if (hasTimeoutTlv && timeout == 0)
```

Reply:
No need to check `processTimeoutTlv` since `timeout` is assigned `config.mMlrTimeout` when  `!processTimeoutTlv`, and `config.mMlrTimeout` must be non-zero since we check that this there must be a PBBR previously.